fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/ExplainTokens.java

```diff
@@ -0,0 +1,1110 @@
+/*
+ * ExplainTokensWithPrecedence.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.google.common.base.Verify;
+import com.google.common.collect.Lists;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.IntBinaryOperator;
+import java.util.function.Supplier;
+
+/**
+ * Class to hold information about a list of tokens. Tokens are created during the explain walk and rendered afterward.
+ * Tokens can be simple, or they can be nested (via {@link NestedToken}). If a token is nested it contains an instance
+ * of this class making this general structure more of a tree than a list.
+ * <br>
+ * Most tokens render into some {@link CharSequence} using an {@link ExplainFormatter}. Some tokens also interact with
+ * the formatter in way that changes the internal state of the formatter. Those tokens when rendered exert a side effect
+ * on the formatter (e.g. declaring an alias, etc.)
+ * <br>
+ * Tokens may render differently for different {@link ExplainLevel}s. For instance, a token that renders some
+ * information in {@link ExplainLevel#ALL_DETAILS} may only render the string {@code ...} in
+ * {@link ExplainLevel#STRUCTURE}.
+ * <br>
+ * It is important to keep track of the number of characters the renderer would produce if rendered for each individual
+ * explain level. Unfortunately, this is not an exact science! While most tokens render to exactly the same character
+ * sequence on each explain level (for e.g. a keyword), for some tokens (e.g. aliases) we just do not know what they
+ * will render into as not all information is available when the token is created during the explain phase. Staying
+ * with the example, we do not know the length of an alias rendering until we know the symbol table to resolve the
+ * alias.
+ * <br>
+ * While we do not know the exact number of characters a token will render into a priori, we can estimate the bounds
```

Comment:
Sorry, this is pedantic:

```suggestion
 * While we do not know the exact number of characters a token will render into <i>a priori</i>, we can estimate the bounds
```

This is following the style suggestion that foreign phrases should be rendered in italics. In other places, we've used `<em>` to render phrases in italics, but that's when using italics for <em>emphasis</em>, whereas this is using italics <i>pro forma</i>, so I think it really should be `<i>`, even though they will render the same.